mitiq/factories.py

```diff
@@ -271,30 +276,33 @@ def static_reduce(instack: List[float], outstack: List[float], asymptote: Union[
             )
 
         # CASE 1: asymptote is None.
-        # TODO: it works, but there must be better way of doing this.
+        # TODO: it works, but there must be a better way of doing this.
         # Note: *args does not seem to work with curve_fit.
         # For the moment only orders up to 3 are suppoerted.
-        def ansatz_zero(x: float, asympt: float, b: float) -> float:
+        def ansatz_zero(_x: float, asympt: float, b: float) -> float:
             """Ansatz function of order 0"""
             return asympt + b
         def ansatz_one(x: float, asympt: float, b: float, z_one: float) -> float:
             """Ansatz function of order 1"""
-            return asympt + b * np.exp(-z_one * x)
+            return asympt + b * np.exp(z_one * x)
         def ansatz_two(x: float, asympt: float, b: float, z_one: float, z_two: float) -> float:
             """Ansatz function of order 2."""
-            return asympt + b * np.exp(-z_one * x - z_two * x ** 2)
+            return asympt + b * np.exp(z_one * x + z_two * x ** 2)
         def ansatz_three(
                 x: float, asympt: float, b: float, z_one: float, z_two: float, z_three: float
             ) -> float:
             """Ansatz function of order 3."""
-            return asympt + b * np.exp(-z_one * x - z_two * x ** 2 - z_three * x ** 3)
+            return asympt + b * np.exp(z_one * x + z_two * x ** 2 + z_three * x ** 3)
```

Comment:
To reduce the number of functions, what about the signature

```python
def ansatz(x: float, a: float, b: float, coeffs: List[float]) -> float:
```

Then, the order is determined and defined by the number of coefficients, i.e., by `len(coeffs)`. The body of the function could be

```python
return a + b * np.exp(np.polyval(coeffs, x))
```

I think [`np.polyval`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.polyval.html) will correctly handle the case when `len(coeffs) = 0` but please double check.

Reply:
Thanks Ryan, I have done more or less as you suggested and now it works! 
Since `curve_fit` wants only scalar variables, I used `*coeffs` as argument. The initialization values `p0` of `curve_fit` determines the `len(coeffs) ` and so `curve_fit` knows how many parameters should be optimized.

def ansatz(x: float, *coeffs: float):
            """Ansatz function of generic order."""
            # Coefficients of the polynomial to be exponentiated
            z_coeffs = coeffs[2:][::-1]
            return coeffs[0] + coeffs[1] * np.exp(x * np.polyval(z_coeffs, x))